lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -463,12 +460,42 @@ struct ConvertMul : public OpConversionPattern<MulOp> {
         AffineMap::get(2, 0, {d1}),      // j
         AffineMap::get(2, 0, {d0 + d1})  // i+j
     };
-
     auto polymulOutput = b.create<arith::ConstantOp>(
         polymulTensorType,
         DenseElementsAttr::get(
             polymulTensorType,
             APInt(polymulTensorType.getElementTypeBitWidth(), 0L)));
+
+    // When our modulus is not a machine-word size (power of 2) we can not rely
+    // on the natural overflow behaviour during the computation of
+    // acc = a[i] * b[j] + c[i+j]. If we compute acc mod cmod then we are able
+    // to ensure that we will not overflow the intermediate type sized integer.
+    // This is because a[i] * b[j] + c[i+j] < cmod^2 + cmod < maximum of
+    // intermediate type. Which lets us not have to compute the modulus after
+    // the multiplication step and the addition step.
+    //
+    // Eg: cmod = 7, then the intermediate type is i6 and 7 + 7^2 = 56 < 64 =
+    // 2^6.
+    //
+    // Further, when we truncate from the intermediate type back to the
+    // polynomial type, we can't rely on the truncation behaviour to ensure that
+    // the value congruent with respect to cmod. So we will have to manually
+    // convert by computing acc + cmod mod cmod.
+    //
+    // Eg: Let acc = -12. Then arith.remsi acc 7 = -5 : i6, and so,
+    // arith.trunci -5 : i6 -> i3 = 3. -5 is not congruent with 3 mod 7. So we
+    // will need to compute -5 + 7 mod 7 = 2.
```

Comment:
The addition should be necessary, otherwise the truncation can cause an error. In the example `-5 = 111011 : i6`, if we directly truncate it to an i3 then we would get `3 = 011 : i3` which are not congruent modulo 7. So we do the addition to ensure that they are in [0,7) before the truncation to avoid such an error.

Reply:
Thanks!